core/src/main/java/io/substrait/relation/physical/HashJoin.java

```diff
@@ -0,0 +1,85 @@
+package io.substrait.relation.physical;
+
+import io.substrait.expression.Expression;
+import io.substrait.expression.FieldReference;
+import io.substrait.proto.HashJoinRel;
+import io.substrait.relation.BiRel;
+import io.substrait.relation.HasExtension;
+import io.substrait.relation.RelVisitor;
+import io.substrait.type.Type;
+import io.substrait.type.TypeCreator;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Stream;
+import org.immutables.value.Value;
+
+@Value.Immutable
+public abstract class HashJoin extends BiRel implements HasExtension {
+
+  public abstract Optional<List<FieldReference>> getLeftKeys();
+
+  public abstract Optional<List<FieldReference>> getRightKeys();
```

Comment:
I would argue that both `getLeftKeys` and `getRightKeys` should be required fields for a HashJoin (i.e. `List<FieldReference>`) as users should be forced to set them. As is, a user can construct a HashJoin with only left keys or only right keys.

Admittedly they can do this in the proto as well as the fields are not marked as required, but we can choose to be stricter in the library to guide their hand. If a user really wants a hash join on no keys, they should set the keys to empty lists.

Reply:
This is a good point. I think I didn't think about it well in the first go. +1 for this change. 
I will add it.

Reply:
I updated the code.